File: Go/problem036.go
package projecteuler

// #36 Double-base palindromes - Project Euler
// http://projecteuler.net/index.php?section=problems&id=36
import (
	"fmt"
	"reflect"
)

func isPalindromicBase(base, x int) bool {
	v1 := []int{}
	v2 := []int{}

	for x > 0 {
		v1 = append(v1, x%base)
		v2 = append([]int{x % base}, v2...)
		x /= base
	}

	return reflect.DeepEqual(v1, v2)
}

func isPalindromic2Base(x int) bool {
	return isPalindromicBase(10, x) && isPalindromicBase(2, x)
}

func P036() {
	ans := 0
	for i := 1; i < 1_000_000; i++ {
		if isPalindromic2Base(i) {
			ans += i
		}
	}
	fmt.Println(ans)
}

File: Zig/problem036.zig
// zig version 0.13.0
const std = @import("std");
const stdout = std.io.getStdOut().writer();
const ArrayList = std.ArrayList;
const allocator = std.heap.page_allocator;

// #36 Double-base Palindromes - Project Euler
// http://projecteuler.net/problem=36
fn isPalindromic(n: u64, base: u64) !bool {
    var digits = ArrayList(u64).init(allocator);
    defer digits.deinit();

    var x = n;
    while (x > 0) : (x = @divTrunc(x, base)) {
        try digits.append(@mod(x, base));
    }

    for (0..@divTrunc(digits.items.len, 2)) |i| {
        if (digits.items[i] != digits.items[digits.items.len - 1 - i]) {
            return false;
        }
    }

    return true;
}

fn isPalindromicInBase10And2(i: u64) !bool {
    return try isPalindromic(i, 10) and try isPalindromic(i, 2);
}

pub fn main() !void {
    var ans: u64 = 0;

    const MAX: u64 = 1000000;
    for (1..MAX) |i| {
        if (try isPalindromicInBase10And2(i)) {
            ans += i;
        }
    }

    try stdout.print("{}\n", .{ans});
}
